test/type-applications/Main.hs

```diff
@@ -331,9 +345,135 @@ propUnifyWildcardRigid = forAllShrink arbitrary shrink $ \(scope, index) ->
+  fnRenamed <- failLeft . runRenameM . renameCompT $ testFn
+  argRenamed <- failLeft . runRenameM . renameValT $ testArg
+  result <-
+    either (assertFailure . show) pure $
+      checkApp tyAppTestDatatypes fnRenamed [Just argRenamed]
+  assertEqual "polyAppMaybe" result (BuiltinFlat IntegerT)
+
+-- Applies a mixture of polymorphic and concrete arguments to `defaultLeft` and checks that the  return
+-- type is what we expected after unification
+polymorphicApplicationE :: TestTree
+polymorphicApplicationE = testCase "polyAppEither" $ do
+  -- a = a' (arbitrary unifiable)
+  -- b = Bool
+  -- c = Integer
+  let arg1 = Abstraction $ Unifiable ix0
+      arg2 = ThunkT (Comp0 $ BuiltinFlat BoolT :--:> ReturnT (BuiltinFlat IntegerT))
+      arg3 = Datatype "Either" . Vector.fromList $ [arg1, BuiltinFlat BoolT]
+  fnRenamed <- failLeft . runRenameM . renameCompT $ defaultLeft
+  actual <-
+    either (assertFailure . show) pure $
+      checkApp tyAppTestDatatypes fnRenamed (pure <$> [arg1, arg2, arg3])
+  assertEqual "polyAppEither" actual (BuiltinFlat IntegerT)
+
+-- Applies a mixture of polymorphic and concrete arguments to `defaultPair` and checks that the return type
+-- is what we expected after unification
+polymorphicApplicationP :: TestTree
+polymorphicApplicationP = testCase "polyAppPair" $ do
+  -- a = a' (arbitrary unifiable)
+  -- b = Bool
+  -- c = Integer
+  let arg1 = Abstraction $ Unifiable ix0
+      arg2 = BuiltinFlat BoolT
+      arg3 = ThunkT $ Comp0 $ Abstraction (Rigid 1 ix0) :--:> BuiltinFlat BoolT :--:> ReturnT (BuiltinFlat IntegerT)
+      arg4 = Datatype "Pair" (Vector.fromList [arg1, BuiltinFlat BoolT])
+  fnRenamed <- failLeft . runRenameM . renameCompT $ defaultPair
+  actual <-
+    either (assertFailure . show) pure $
+      checkApp tyAppTestDatatypes fnRenamed (pure <$> [arg1, arg2, arg3, arg4])
+  assertEqual "polyAppPair" actual (BuiltinFlat IntegerT)
+
+-- Checks whether `forall a. Maybe a -> Integer` unifies properly with `Maybe Bool -> Integer`
+unifyMaybe :: TestTree
+unifyMaybe = testCase "unifyMaybe" $ do
+  let testFn =
+        Comp1 $
+          Datatype "Maybe" (Vector.fromList [tyvar Z ix0])
+            :--:> ReturnT (BuiltinFlat IntegerT)
+      testArg = Datatype "Maybe" (Vector.fromList [BuiltinFlat BoolT])
+  fnRenamed <- failLeft . runRenameM . renameCompT $ testFn
+  result <-
+    either (assertFailure . catchInsufficientArgs) pure $
+      checkApp tyAppTestDatatypes fnRenamed [Just testArg]
+  assertEqual "unifyMaybe" result (BuiltinFlat IntegerT)
+  where
+    catchInsufficientArgs :: TypeAppError -> String
+    catchInsufficientArgs = \case
+      InsufficientArgs _ fn _ -> prettyStr fn
+      other -> show other
+
+-- Tests whether renaming works as "expected" with free type variables.
+-- NOTE/FIXME/REVIEW: This test *probably* shouldn't pass, since it lets a free type variable through
+--                    by converting it into a rigid where it should fail.
```

Comment:
This test probably shouldn't be here either. I'd either remove it or move it to the renaming suite.